File: Deno/2024/2024-004/2024-004.ts
// Read the grid from the input file
const readGrid = async (filePath) => {
  const data = await Deno.readTextFile(filePath);
  return data.trim().split("\n").map((line) => line.split(""));
};

// Define the target word and movement directions
const TARGET_WORD = "XMAS";
const DIRECTIONS = [
  [0, 1],   // Right
  [1, 0],   // Down
  [1, 1],   // Diagonal-right-down
  [1, -1],  // Diagonal-left-down
  [0, -1],  // Left
  [-1, 0],  // Up
  [-1, -1], // Diagonal-left-up
  [-1, 1],  // Diagonal-right-up
];

// Function to check if the word exists in a given direction
const checkWord = (grid, word, x, y, dx, dy) => {
  const rows = grid.length;
  const cols = grid[0].length;
  for (let i = 0; i < word.length; i++) {
    const nx = x + i * dx;
    const ny = y + i * dy;
    if (nx < 0 || ny < 0 || nx >= rows || ny >= cols || grid[nx][ny] !== word[i]) {
      return false;
    }
  }
  return true;
};

// Function to count all occurrences of the target word
const countOccurrences = (grid, word) => {
  let count = 0;
  const rows = grid.length;
  const cols = grid[0].length;
  for (let x = 0; x < rows; x++) {
    for (let y = 0; y < cols; y++) {
      for (const [dx, dy] of DIRECTIONS) {
        if (checkWord(grid, word, x, y, dx, dy)) {
          count++;
        }
      }
    }
  }
  return count;
};

// Function to count all X-MAS patterns
const countXmasPatterns = (grid) => {
  let count = 0;
  const rows = grid.length;
  const cols = grid[0].length;
  for (let x = 1; x < rows - 1; x++) {
    for (let y = 1; y < cols - 1; y++) {
      const center = grid[x][y];
      const topLeft = grid[x - 1][y - 1];
      const topRight = grid[x - 1][y + 1];
      const bottomLeft = grid[x + 1][y - 1];
      const bottomRight = grid[x + 1][y + 1];

      if (
        center === "A" &&
        (
          (topLeft === "M" && topRight === "S" && bottomLeft === "M" && bottomRight === "S") ||
          (topLeft === "S" && topRight === "M" && bottomLeft === "S" && bottomRight === "M") ||
          (topLeft === "M" && topRight === "M" && bottomLeft === "S" && bottomRight === "S") ||
          (topLeft === "S" && topRight === "S" && bottomLeft === "M" && bottomRight === "M")
        )
      ) {
        count++;
      }
    }
  }
  return count;
};

// Main execution
const main = async () => {
  const filePath = "/uploads/input.txt";
  const grid = await readGrid(filePath);
  console.log("Count of XMAS occurrences:", countOccurrences(grid, TARGET_WORD));
  console.log("Total XMAS patterns:", countXmasPatterns(grid));
};

main();

File: F#-002/2024/2024-004/2024-004.fs
open System
open System.IO

// Read the grid from the input file
let readGrid filePath =
    File.ReadAllLines(filePath)
    |> Array.map (fun line -> line.ToCharArray())

// Define the target word and movement directions
let targetWord = "XMAS"
let directions = [
    (0, 1)   // Right
    (1, 0)   // Down
    (1, 1)   // Diagonal-right-down
    (1, -1)  // Diagonal-left-down
    (0, -1)  // Left
    (-1, 0)  // Up
    (-1, -1) // Diagonal-left-up
    (-1, 1)  // Diagonal-right-up
]

// Function to check if the word exists in a given direction
let checkWord grid (word: string) x y dx dy =
    let rows = Array.length grid
    let cols = Array.length grid.[0]
    let wordLength = word.Length
    let isValid = 
        [0 .. wordLength - 1]
        |> List.forall (fun i ->
            let nx = x + i * dx
            let ny = y + i * dy
            nx >= 0 && ny >= 0 && nx < rows && ny < cols && grid.[nx].[ny] = word.[i]
        )
    isValid

// Count all occurrences of the target word
let countOccurrences grid word =
    let rows = Array.length grid
    let cols = Array.length grid.[0]
    [ for x in 0 .. rows - 1 do
        for y in 0 .. cols - 1 do
            for (dx, dy) in directions do
                if checkWord grid word x y dx dy then yield 1 ]
    |> List.sum

// Count all XMAS patterns
let countXmasPatterns grid =
    let rows = Array.length grid
    let cols = Array.length grid.[0]
    [ for x in 1 .. rows - 2 do
        for y in 1 .. cols - 2 do
            let center = grid.[x].[y]
            let topLeft = grid.[x - 1].[y - 1]
            let topRight = grid.[x - 1].[y + 1]
            let bottomLeft = grid.[x + 1].[y - 1]
            let bottomRight = grid.[x + 1].[y + 1]
            if center = 'A' && (
                (topLeft = 'M' && topRight = 'S' && bottomLeft = 'M' && bottomRight = 'S') ||
                (topLeft = 'S' && topRight = 'M' && bottomLeft = 'S' && bottomRight = 'M') ||
                (topLeft = 'M' && topRight = 'M' && bottomLeft = 'S' && bottomRight = 'S') ||
                (topLeft = 'S' && topRight = 'S' && bottomLeft = 'M' && bottomRight = 'M')
            ) then yield 1 ]
    |> List.sum

// Main function
[<EntryPoint>]
let main argv =
    let filePath = "/uploads/input.txt"
    let grid = readGrid filePath
    let xmasCount = countOccurrences grid targetWord
    printfn "Count of XMAS occurrences: %d" xmasCount
    let xmasPatterns = countXmasPatterns grid
    printfn "Total XMAS patterns: %d" xmasPatterns
    0
